openedx/core/djangoapps/discussions/docs/decisions/0002-organization-wide-provider-filtering.rst

```diff
@@ -0,0 +1,155 @@
+      an engineer to make a code/config change and pull request
+- Database-backed
+  - pros:
+    - ease of use
+      - by (django) admins
+  - cons:
+    - possible schema lock-in
+      - eg: do we scope it to `organizations` only?
+            or do we generalize it enough to use elsewhere?
+
+
+Decision
+--------
+
+We propose to implement this as a Django database-backed model,
+`ProviderFilter`.
+
+The underlying model can be represented as:
+
+.. code-block:: python
+    class ProviderFilter(StackedConfigurationModel):
+        allow = models.CharField(blank=True, ...)
+        deny = models.CharField(blank=True, ...)
+
+- `allow` is a space-delimited string of providers to allow
+
+- `deny` is a space-delimited string of providers to deny
+
+By extending the `StackedConfigurationModel`, we gain the flexibility of
+stacking global/site/org/course/run values to not only provide
+organization-wide filtering today, but to arbitrary levels of
+granularity in the future. As this base model already exists in
+`edx-platform`, integration is low-effort.
+
+
+Examples
+--------
+
+With this system, we can configure global/site-wide defaults,
+
+.. code-block:: python
+    ProviderFilter.objects.create()
+
+And then override to deny an external provider for an organization:
+
+.. code-block:: python
+    ProviderFilter.objects.create(
+        org='Piazza-Less',
+        deny='lti-providerx',
+    )
+
+Or override to deny _all_ external providers for an organization:
+
+.. code-block:: python
+    ProviderFilter.objects.create(
+        org='InternalOrganization',
+        allow='cs_comments_service',
+    )
+
+And grant an exemption to a specific course:
+
+.. code-block:: python
+    ProviderFilter.objects.create(
+        # a course in 'InternalOrganization'
+        course=course,
+        allow='cs_comments_service lti-providerx',
+    )
+
+
+Logic
+-----
+
+.. code-block:: python
+    _filter = get_filter(course_key)
+    if _filter is None:
+        allow = set()
+        deny = set()
+    available = defaults
+    if len(allow) > 0:
+        available = available.union(allow)
+    if len(deny) > 0:
+        available = available.difference(deny)
+
+By default, all installed providers are available.
+
+If no database record exists for the organization, the allow and deny
+lists are considered empty, ie, there are no restrictions.
+
+If there are any items in the allow list, all _other_ providers will be
+removed from the available list.
+
+If there are any items in the deny list, _these_ providers will be
+removed from the available list.
+
+The order of operations between the deny and allow should be
+interchangeable.
+
+
+Discussion
+----------
+
+Is there a best (or existing?) practice for storing a list of strings in
+the database via django?
+I've seen some custom fields googling around, but for right now, I'm
+just storing a string a splitting/joining it via a helper :shrug:
```

Comment:
Just curious how we will know the canonical names of the service providers when we add them to allow/deny lists? also is it common to do space-separated strings over comma separated? (the only config I recall doing (darklangconfig) I used comma-separated, and did have to manually .split the TextField)

Reply:
In the short term, the names are hard-coded strings in a dict, basically, so the only way to know 'em is to read the code... in the long run, though, I expect we'll replace our hard-coded list with a database backed "app provider" model of some sort... @stvstnfrd, how do you think this system would change to accommodate that?  I'd expect we'd want to use foreign keys then, rather than a string.  Do you think it'd be an easy migration?

Reply:
I think one way to simplify things is to just list all the available providers in a read-only field in the admin panel. Admins can then copy-paste from this list of available providers. There can also be a list called "filtered providers" which lists the providers post-filtering, so admin's know how it's working.